src/DSi_DSP.cpp

```diff
@@ -0,0 +1,592 @@
+
+    if ( PDATAReadFifo.IsFull ()) r |= 1<<5;
+    if (!PDATAReadFifo.IsEmpty()) r |=(1<<6)|(1<<0);
+
+    if (!TeakraCore->SendDataIsEmpty(0)) r |= 1<<13;
+    if (!TeakraCore->SendDataIsEmpty(1)) r |= 1<<14;
+    if (!TeakraCore->SendDataIsEmpty(2)) r |= 1<<15;
+    if ( TeakraCore->RecvDataIsReady(0)) r |= 1<<10;
+    if ( TeakraCore->RecvDataIsReady(1)) r |= 1<<11;
+    if ( TeakraCore->RecvDataIsReady(2)) r |= 1<<12;
+
+    return r;
+}
+
+void IrqRep0()
+{
+    if (DSP_PCFG & (1<< 9)) NDS::SetIRQ(0, NDS::IRQ_DSi_DSP);
+}
+void IrqRep1()
+{
+    if (DSP_PCFG & (1<<10)) NDS::SetIRQ(0, NDS::IRQ_DSi_DSP);
+}
+void IrqRep2()
+{
+    if (DSP_PCFG & (1<<11)) NDS::SetIRQ(0, NDS::IRQ_DSi_DSP);
+}
+void IrqSem()
+{
+    DSP_PSTS |= 1<<9;
+    // apparently these are always fired?
+    NDS::SetIRQ(0, NDS::IRQ_DSi_DSP);
+}
+
+void AudioCb(std::array<s16, 2> frame)
+{
+    // TODO
+}
+
+bool Init()
+{
+    TeakraCore = new Teakra::Teakra();
+    SCFG_RST = false;
+
+    if (!TeakraCore) return false;
+
+    TeakraCore->SetRecvDataHandler(0, IrqRep0);
+    TeakraCore->SetRecvDataHandler(1, IrqRep1);
+    TeakraCore->SetRecvDataHandler(2, IrqRep2);
+
+    TeakraCore->SetSemaphoreHandler(IrqSem);
+
+    // these happen instantaneously and without too much regard for bus aribtration
+    // rules, so, this might have to be changed later on
+    Teakra::AHBMCallback cb;
+    cb.read8 = DSi::ARM9Read8;
+    cb.write8 = DSi::ARM9Write8;
+    cb.read16 = DSi::ARM9Read16;
+    cb.write16 = DSi::ARM9Write16;
+    cb.read32 = DSi::ARM9Read32;
+    cb.write32 = DSi::ARM9Write32;
+    TeakraCore->SetAHBMCallback(cb);
+
+    TeakraCore->SetAudioCallback(AudioCb);
+
+    //PDATAReadFifo = new FIFO<u16>(16);
+    //PDATAWriteFifo = new FIFO<u16>(16);
+
+    return true;
+}
+void DeInit()
+{
+    //if (PDATAWriteFifo) delete PDATAWriteFifo;
+    if (TeakraCore) delete TeakraCore;
+
+    //PDATAReadFifo = NULL;
+    //PDATAWriteFifo = NULL;
+    TeakraCore = NULL;
+}
+
+void Reset()
+{
+    DSPTimestamp = 0;
+
+    DSP_PADR = 0;
+    DSP_PCFG = 0;
+    DSP_PSTS = 0;
+    DSP_PSEM = 0;
+    DSP_PMASK = 0xff;
+    DSP_PCLEAR = 0;
+    DSP_CMD[2] = DSP_CMD[1] = DSP_CMD[0] = 0;
+    DSP_REP[2] = DSP_REP[1] = DSP_REP[0] = 0;
+    PDataDMALen = 0;
+
+    PDATAReadFifo.Clear();
+    //PDATAWriteFifo->Clear();
+    TeakraCore->Reset();
+
+    NDS::CancelEvent(NDS::Event_DSi_DSP);
+}
+
+bool IsRstReleased()
+{
+    return SCFG_RST;
+}
+void SetRstLine(bool release)
+{
+    SCFG_RST = release;
+    Reset();
+    DSPTimestamp = NDS::ARM9Timestamp; // only start now!
+}
+
+void OnMBKCfg(char bank, u32 slot, u8 oldcfg, u8 newcfg, u8* nwrambacking)
+{
+    if (bank != 'B' && bank != 'C')
+    {
+        printf("WTF?? (DSP MBK recfg, nonsense bank '%c')\n", bank);
+        return;
+    }
+
+    bool olddsp = (oldcfg & 3) >= 2, // was mapped to the DSP
+         newdsp = (newcfg & 3) >= 2; // will be mapped to the DSP
+
+    // nothing changes
+    if (olddsp == newdsp)
```

Comment:
(note to self) This code forgets about the case of mapping the same NWRAM bank differently to the DSP (if that's a possibility, should be investigated), and thus needs to be reworked a bit.